src/gerbil/runtime/c3.ss

```diff
@@ -0,0 +1,96 @@
+;;; -*- Gerbil -*-
+;;; © fare
+;;; The C3 Linearization algorithm to topologically sort an inheritance DAG
+;;; into a precedence list such that:
+;;;  1. The precedence list is a linearization of the inheritance DAG
+;;;  2. Every precedence list includes as a sublist the precedence list of every super
+;;;  3. The list of supers is a sublist of the precedence list
+;;;  4. The elements that appear earlier in the list of precedence lists followed by
+;;;     the super list appear earlier in the precedence list.
+;;;     (This heuristic choice of an order makes the linearization deterministic).
+;;; In the above, sublist means that the super list includes all the sublist elements
+;;; in the same order but not necessarily consecutively.
+;;;
+;;; C3 has since been adopted for multiple inheritance by many modern object systems
+;;; for its ordering consistency properties:
+;;; OpenDylan, Python, Raku, Parrot, Solidity, PGF/TikZ.
+;;;       https://en.wikipedia.org/wiki/C3_linearization
+;;;       https://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.19.3910
+;;;
+;;; See tests in ../test/c3-test.ss
+
+prelude: :gerbil/core
+package: gerbil/runtime
+namespace: #f
+
+(export #t)
+
+;; C3 linearization algorithm: given a top object x from which to compute the precedence list,
+;; - rhead is [x] or [] depending on whether to include x as head of the result.
+;; - supers is (get-supers x), the list of direct supers of x
+;; - get-precedence-list gets the precedence list for a super s, including s itself in front
+;; - eqpred is an equality predicate between list elements
+;;   (passed around because somehow equal? doesn't work correctly on runtime type-descriptors).
+;; - get-name gets the name of a object/class, for debugging only.
+;; : (List X) (List X) (X -> (NonEmptyList X)) ?(X X -> Bool) ?(X -> Y) -> (List X)
+(def (c3-linearize rhead supers get-precedence-list (eqpred eqv?) (get-name identity))
+  (def super-precedence-lists (map get-precedence-list supers)) ;; : (List (NonEmptyList X))
+  (c3-linearize-loop rhead (remove-nulls (append super-precedence-lists [supers])) eqpred get-name))
```

Comment:
this is a rather inefficient way to do it, too many intermediate lists.

Reply:
Nah, actually it's not too bad: most of these lists are just being shared and walked. There's just the list of heads that could be destructively / linearly updated rather than reconstructed all the time, but doing it that way is messy for a small gain on an operation that shouldn't be too time-critical. We can always super-optimize later if this ever becomes a bottleneck.

Reply:
But OK, I removed a lot of consing for you.